spock-core/src/main/java/org/spockframework/runtime/extension/builtin/RetryIterationInterceptor.java

```diff
@@ -74,26 +77,32 @@ public void intercept(IMethodInvocation invocation) throws Throwable {
   }
 
   private class IterationState {
-    private final ThreadLocal<Boolean> finalIteration = ThreadLocal.withInitial(() -> false);
-    private final ThreadLocal<List<Throwable>> throwables = new ThreadLocal<>();
-
-    void startIteration() {
-      throwables.set(new ArrayList<>());
+    private final Map<IterationInfo, AtomicBoolean> finalIteration = new ConcurrentHashMap<>();
+    private final Map<IterationInfo, List<Throwable>> throwables = new ConcurrentHashMap<>();
+
+    void resetFailures(IterationInfo iteration) {
+      this.throwables.computeIfPresent(iteration, (key, throwables) -> {
+        throwables.clear();
+        return throwables;
+      });
```

Comment:
I find it hard to understand that the field and the lamda parameter have the same name.
```suggestion
      this.throwables.computeIfPresent(iteration, (key, throwablesList) -> {
        throwablesList.clear();
        return throwablesList;
      });
```